// mjml-hero/lib/Hero.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _mjmlCore = require('mjml-core');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var tagName = 'mj-hero';
var parentTag = ['mj-container', 'mj-wrapper'];
var defaultMJMLDefinition = {
  attributes: {
    'mode': 'fixed-height',
    'height': '0px',
    'background-url': null,
    'background-width': '0px',
    'background-height': '0px',
    'background-position': 'center center',
    'padding': '0px',
    'padding-bottom': null,
    'padding-left': null,
    'padding-right': null,
    'padding-top': null,
    'background-color': '#ffffff',
    'vertical-align': 'top'
  }
};

var endingTag = false;
var baseStyles = {
  div: {
    margin: '0 auto'
  },
  table: {
    width: '100%'
  },
  tr: {
    verticalAlign: 'top'
  },
  edge: {
    paddingBottom: '0%',
    width: '0.01%',
    msoPaddingBottomAlt: 0
  },
  hero: {
    backgroundRepeat: 'no-repeat',
    verticalAlign: 'top',
    backgroundSize: 'cover'
  }
};

var postRender = function postRender($) {
  var backgroundCropTop = 0;
  var backgroundCropBottom = 0;
  var dataCrop = '';
  var $element = void 0;
  var backgroundWidth = 0;
  var backgroundHeight = 0;
  var backgroundUrl = '';

  $('.mj-hero-outlook').each(function () {
    backgroundUrl = $(this).data('background-url');
    backgroundWidth = $(this).data('background-width');
    backgroundHeight = $(this).data('background-height');
    backgroundCropTop = 0;
    backgroundCropBottom = 0;
    dataCrop = '';

    $element = $(this).find('.mj-hero-fixed-height').first();

    if ($element.length) {
      dataCrop = $element.data('crop');
      backgroundHeight = $element.data('background-height');

      backgroundCropTop = dataCrop.split(', ')[0].split(':')[1];
      backgroundCropBottom = dataCrop.split(', ')[1].split(':')[1];

      $element.removeAttr('class').removeAttr('data-background-height').removeAttr('data-crop');
    }

    backgroundWidth = parseInt(backgroundWidth.replace('px', '')) * 0.75;
    backgroundHeight = parseInt(backgroundHeight.replace('px', '')) * 0.75;

    $(this).before(_mjmlCore.helpers.startConditionalTag + '\n          <v:image\n            xmlns:v="urn:schemas-microsoft-com:vml"\n            croptop="' + backgroundCropTop + '"\n            cropbottom="' + backgroundCropBottom + '"\n            style="width:' + backgroundWidth + 'pt;height:' + backgroundHeight + 'pt;position:absolute;top:0;mso-position-horizontal:center;border:0;z-index:-3;"\n            src="' + backgroundUrl + '" />\n        ' + _mjmlCore.helpers.endConditionalTag).removeClass('mj-hero-outlook').removeAttr('data-background-width').removeAttr('data-background-height').removeAttr('data-background-url');
  });

  return $;
};

var Hero = (0, _mjmlCore.MJMLElement)(_class = function (_Component) {
  _inherits(Hero, _Component);

  function Hero() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Hero);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Hero.__proto__ || Object.getPrototypeOf(Hero)).call.apply(_ref, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Hero, [{
    key: 'getFixedHeight',
    value: function getFixedHeight() {
      var _props = this.props,
          mjAttribute = _props.mjAttribute,
          getPadding = _props.getPadding;

      var height = 0;
      var paddingTop = 0;
      var paddingBottom = 0;

      height = parseInt(mjAttribute('height').replace('px', ''));
      paddingTop = Math.abs(getPadding('top'));
      paddingBottom = Math.abs(getPadding('bottom'));

      return height - paddingTop - paddingBottom;
    }
  }, {
    key: 'getBackgroundCrop',
    value: function getBackgroundCrop() {
      var mjAttribute = this.props.mjAttribute;

      var height = parseInt(mjAttribute('height').replace('px', ''));
      var backgroundHeight = parseInt(mjAttribute('background-height').replace('px', ''));
      var backgroundPositionTop = mjAttribute('background-position').trim().split(' ')[0] || 'center';
      var cropTop = 0;
      var cropBottom = 0;

      if (height < backgroundHeight) {
        switch (backgroundPositionTop) {
          case 'top':
            cropBottom = Math.round((backgroundHeight - height) / backgroundHeight * 100) / 100;
            break;
          case 'center':
            cropTop = Math.round((backgroundHeight - height) / 2 / backgroundHeight * 100) / 100;
            cropBottom = cropTop;
            break;
          case 'bottom':
            cropTop = Math.round((backgroundHeight - height) / backgroundHeight * 100) / 100;
            break;
        }
      }

      return 'croptop:' + cropTop + ', cropbottom:' + cropBottom;
    }
  }, {
    key: 'getBackgroundRatio',
    value: function getBackgroundRatio() {
      var mjAttribute = this.props.mjAttribute;

      var backgroundWidth = parseInt(mjAttribute('background-width').replace('px', ''));
      var backgroundHeight = parseInt(mjAttribute('background-height').replace('px', ''));

      return Math.round(backgroundHeight / backgroundWidth * 100 * 10000) / 10000;
    }
  }, {
    key: 'getBackgroundStyle',
    value: function getBackgroundStyle() {
      var mjAttribute = this.props.mjAttribute;

      var background = '';

      background += (mjAttribute('background-color') || '') + ' ';
      background += mjAttribute('background-url') ? 'url(' + mjAttribute('background-url') + ') ' : '';
      background += 'no-repeat ';
      background += mjAttribute('background-position') + ' ';
      background += '/ cover';

      return background;
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      var _props2 = this.props,
          mjAttribute = _props2.mjAttribute,
          getPadding = _props2.getPadding,
          defaultUnit = _props2.defaultUnit,
          parentWidth = _props2.parentWidth;

      var backgroundRatio = this.getBackgroundRatio();
      var backgroundStyle = this.getBackgroundStyle();
      var width = parentWidth;

      if (mjAttribute('width')) {
        width = mjAttribute('width');
      }

      return _mjmlCore.helpers.merge({}, baseStyles, {
        div: {
          maxWidth: defaultUnit(width, 'px')
        },
        edge: {
          paddingBottom: backgroundRatio + '%'
        },
        hero: {
          background: backgroundStyle,
          paddingTop: getPadding('top'),
          paddingLeft: getPadding('left'),
          paddingRight: getPadding('right'),
          paddingBottom: getPadding('bottom'),
          backgroundPosition: mjAttribute('background-position'),
          verticalAlign: mjAttribute('vertical-align')
        }
      });
    }
  }, {
    key: 'isFixedHeight',
    value: function isFixedHeight() {
      var mjAttribute = this.props.mjAttribute;


      return mjAttribute('mode') == 'fixed-height';
    }
  }, {
    key: 'renderFixedHeight',
    value: function renderFixedHeight() {
      var _props3 = this.props,
          mjAttribute = _props3.mjAttribute,
          children = _props3.children,
          defaultUnit = _props3.defaultUnit;


      return _react2.default.createElement(
        'tr',
        { style: this.styles.tr },
        _react2.default.createElement(
          'td',
          {
            className: 'mj-hero-fixed-height',
            height: this.getFixedHeight(),
            'data-background-height': defaultUnit(mjAttribute('height'), 'px'),
            'data-legacy-background': mjAttribute('background-url'),
            'data-crop': this.getBackgroundCrop(),
            style: this.styles.hero },
          children
        )
      );
    }
  }, {
    key: 'renderFluidHeight',
    value: function renderFluidHeight() {
      var _props4 = this.props,
          mjAttribute = _props4.mjAttribute,
          children = _props4.children;


      return _react2.default.createElement(
        'tr',
        { style: this.styles.tr },
        _react2.default.createElement('td', { style: this.styles.edge }),
        _react2.default.createElement(
          'td',
          {
            'data-legacy-background': mjAttribute('background-url'),
            style: this.styles.hero },
          children
        ),
        _react2.default.createElement('td', { style: this.styles.edge })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props5 = this.props,
          mjAttribute = _props5.mjAttribute,
          defaultUnit = _props5.defaultUnit;

      var divClasses = (0, _classnames2.default)('mj-hero-outlook', mjAttribute('css-class'));

      return _react2.default.createElement(
        'div',
        {
          className: divClasses,
          'data-background-width': defaultUnit(mjAttribute('background-width'), 'px'),
          'data-background-height': defaultUnit(mjAttribute('background-height'), 'px'),
          'data-background-url': mjAttribute('background-url'),
          style: this.styles.div },
        _react2.default.createElement(
          'table',
          {
            role: 'presentation',
            cellPadding: '0',
            cellSpacing: '0',
            style: this.styles.table },
          _react2.default.createElement(
            'tbody',
            null,
            this.isFixedHeight() ? this.renderFixedHeight() : this.renderFluidHeight()
          )
        )
      );
    }
  }]);

  return Hero;
}(_react.Component)) || _class;

Hero.tagName = tagName;
Hero.defaultMJMLDefinition = defaultMJMLDefinition;
Hero.endingTag = endingTag;
Hero.baseStyles = baseStyles;
Hero.postRender = postRender;
Hero.parentTag = parentTag;

exports.default = Hero;
