// mjml-invoice/lib/InvoiceItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _mjmlCore = require('mjml-core');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var tagName = 'mj-invoice-item';
var parentTag = ['mj-invoice'];
var endingTag = true;
var defaultMJMLDefinition = {
  attributes: {
    'color': '#747474',
    'font-family': 'Roboto, Ubuntu, Helvetica, Arial, sans-serif',
    'font-size': '14px',
    'name': '',
    'padding': '10px 20px',
    'price': '0',
    'quantity': '0',
    'text-align': 'left'
  }
};
var baseStyles = {
  td: {
    fontWeight: '500',
    lineHeight: '1'
  },
  name: {
    wordBreak: 'break-all'
  },
  quantity: {
    textAlign: 'right'
  }
};

var InvoiceItem = (0, _mjmlCore.MJMLElement)(_class = function (_Component) {
  _inherits(InvoiceItem, _Component);

  function InvoiceItem() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, InvoiceItem);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = InvoiceItem.__proto__ || Object.getPrototypeOf(InvoiceItem)).call.apply(_ref, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(InvoiceItem, [{
    key: 'getStyles',
    value: function getStyles() {
      var _props = this.props,
          mjAttribute = _props.mjAttribute,
          defaultUnit = _props.defaultUnit;


      var styles = _mjmlCore.helpers.merge({}, baseStyles, {
        td: {
          color: mjAttribute('color'),
          fontFamily: mjAttribute('font-family'),
          fontSize: defaultUnit(mjAttribute('font-size')),
          padding: defaultUnit(mjAttribute('padding')),
          textAlign: mjAttribute('text-align')
        }
      });

      styles.name = _mjmlCore.helpers.merge({}, styles.td, styles.name);
      styles.quantity = _mjmlCore.helpers.merge({}, styles.td, styles.quantity);

      return styles;
    }
  }, {
    key: 'render',
    value: function render() {
      var mjAttribute = this.props.mjAttribute;


      return _react2.default.createElement(
        'tr',
        { className: mjAttribute('css-class') },
        _react2.default.createElement(
          'td',
          { style: this.styles.name },
          mjAttribute('name')
        ),
        _react2.default.createElement(
          'td',
          { style: this.styles.td },
          mjAttribute('price')
        ),
        _react2.default.createElement(
          'td',
          { style: this.styles.quantity },
          mjAttribute('quantity')
        )
      );
    }
  }]);

  return InvoiceItem;
}(_react.Component)) || _class;

InvoiceItem.tagName = tagName;
InvoiceItem.parentTag = parentTag;
InvoiceItem.endingTag = endingTag;
InvoiceItem.defaultMJMLDefinition = defaultMJMLDefinition;
InvoiceItem.baseStyles = baseStyles;

exports.default = InvoiceItem;
