// mjml-social/lib/Social.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createComponent = require('mjml-core/lib/createComponent');

exports.default = (0, _createComponent.createBodyComponent)('mj-social', {
  allowedAttributes: {
    'align': 'enum(left,right,center)',
    'border-radius': 'unit(px)',
    'container-background-color': 'color',
    'color': 'color',
    'font-family': 'string',
    'font-size': 'unit(px,%)',
    'font-style': 'string',
    'font-weight': 'string',
    'inner-padding': 'unit(px,%)',
    'line-height': 'unit(px,%)',
    'mode': 'enum(horizontal,vertical)',
    'padding-bottom': 'unit(px,%)',
    'padding-left': 'unit(px,%)',
    'padding-right': 'unit(px,%)',
    'padding-top': 'unit(px,%)',
    'padding': 'unit(px,%){1,4}',
    'table-layout': 'enum(auto,fixed)',
    'vertical-align': 'enum(top,bottom,middle)'
  },
  defaultAttributes: {
    'align': 'center',
    'border-radius': '3px',
    'color': '#333333',
    'font-family': 'Ubuntu, Helvetica, Arial, sans-serif',
    'font-size': '13px',
    'icon-size': '20px',
    'inner-padding': null,
    'line-height': '22px',
    'mode': 'horizontal',
    'padding': '10px 25px',
    'text-decoration': 'none'
  },
  getStyles: function getStyles() {
    return {
      tableVertical: {
        'margin': '0px'
      },
      tableHorizontal: {
        float: 'none',
        display: 'inline-table'
      }
    };
  },
  getSocialElementAttributes: function getSocialElementAttributes() {
    var _this = this;

    return ['border-radius', 'color', 'font-family', 'font-size', 'font-weight', 'icon-size', 'inner-padding', 'line-height'].reduce(function (res, attr) {
      res[attr] = _this.getMjAttribute(attr);
      return res;
    }, {});
  },
  renderHorizontal: function renderHorizontal() {
    var _this2 = this;

    var children = this.props.children;


    return '\n     <!--[if mso | IE]>\n      <table\n        ' + this.generateHtmlAttributes({
      align: this.getMjAttribute('align'),
      border: '0',
      cellpadding: '0',
      cellspacing: '0',
      role: 'presentation'
    }) + '\n      >\n        <tr>\n      <![endif]-->\n      ' + this.renderChildren(children, {
      attributes: this.getSocialElementAttributes(),
      renderer: function renderer(component) {
        return '\n            <!--[if mso | IE]>\n              <td>\n            <![endif]-->\n              <table\n                ' + _this2.generateHtmlAttributes({
          align: _this2.getMjAttribute('align'),
          border: '0',
          cellpadding: '0',
          cellspacing: '0',
          role: 'presentation',
          style: 'tableHorizontal'
        }) + '\n              >\n                ' + component.render() + '\n              </table>\n            <!--[if mso | IE]>\n              </td>\n            <![endif]-->\n          ';
      }
    }) + '\n      <!--[if mso | IE]>\n          </tr>\n        </table>\n      <![endif]-->\n    ';
  },
  renderVertical: function renderVertical() {
    var children = this.props.children;


    return '\n      <table\n        ' + this.generateHtmlAttributes({
      border: '0',
      cellpadding: '0',
      cellspacing: '0',
      role: 'presentation',
      style: 'tableVertical'
    }) + '\n      >\n        ' + this.renderChildren(children, { attributes: this.getSocialElementAttributes() }) + '\n      </table>\n    ';
  },
  render: function render() {
    return '\n      ' + (this.getMjAttribute('mode') == 'horizontal' ? this.renderHorizontal() : this.renderVertical()) + '\n    ';
  }
});
module.exports = exports['default'];
