// mjml-social/lib/SocialElement.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createComponent = require('mjml-core/lib/createComponent');

var defaultSocialNetworks = {
  facebook: {
    'share-url': 'https://www.facebook.com/sharer/sharer.php?u=[[URL]]',
    'background-color': '#3b5998',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/facebook.png'
  },
  twitter: {
    'share-url': 'https://twitter.com/home?status=[[URL]]',
    'background-color': '#55acee',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/twitter.png'
  },
  google: {
    'share-url': 'https://plus.google.com/share?url=[[URL]]',
    'background-color': '#dc4e41',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/google-plus.png'
  },
  pinterest: {
    'share-url': 'https://pinterest.com/pin/create/button/?url=[[URL]]&media=&description=',
    'background-color': '#bd081c',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/pinterest.png'
  },
  linkedin: {
    'share-url': 'https://www.linkedin.com/shareArticle?mini=true&url=[[URL]]&title=&summary=&source=',
    'background-color': '#0077b5',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/linkedin.png'
  },
  instagram: {
    'background-color': '#3f729b',
    'src': 'https://www.mailjet.com/images/theme/v1/icons/ico-social/instagram.png'
  }
};

exports.default = (0, _createComponent.createBodyComponent)('mj-social-element', {
  endingTag: true,
  allowedAttributes: {
    'align': 'enum(left,center,right)',
    'color': 'color',
    'border-radius': 'unit(px)',
    'font-family': 'string',
    'font-size': 'unit(px,%)',
    'font-style': 'string',
    'font-weight': 'string',
    'line-height': 'unit(px,%)',
    'name': 'string',
    'padding-bottom': 'unit(px,%)',
    'padding-left': 'unit(px,%)',
    'padding-right': 'unit(px,%)',
    'padding-top': 'unit(px,%)',
    'padding': 'unit(px,%){1,4}',
    'text-decoration': 'string',
    'width': 'integer'
  },
  defaultAttributes: {
    'align': 'left',
    'color': '#000',
    'border-radius': '3px',
    'font-family': 'Ubuntu, Helvetica, Arial, sans-serif',
    'font-size': '13px',
    'line-height': '22px',
    'padding': '10px 25px',
    'text-decoration': 'none',
    'width': '100%'
  },
  getStyles: function getStyles() {
    var _getSocialAttributes = this.getSocialAttributes(),
        iconSize = _getSocialAttributes['icon-size'],
        backgroundColor = _getSocialAttributes['background-color'];

    return {
      td: {
        'padding': this.getMjAttribute('inner-padding')
      },
      table: {
        'background': backgroundColor,
        'border-radius': this.getMjAttribute('border-radius'),
        'width': iconSize
      },
      icon: {
        'width': iconSize,
        'height': iconSize
      },
      img: {
        'border-radius': this.getMjAttribute('border-radius')
      },
      text: {
        'color': this.getMjAttribute('color'),
        'font-size': this.getMjAttribute('font-size'),
        'font-family': this.getMjAttribute('font-family'),
        'line-height': this.getMjAttribute('line-height'),
        'text-decoration': this.getMjAttribute('text-decoration')
      }
    };
  },
  getSocialAttributes: function getSocialAttributes() {
    var _this = this;

    var socialNetwork = _extends({}, defaultSocialNetworks[this.getMjAttribute('name')]);

    if (socialNetwork['share-url']) {
      socialNetwork['href'] = socialNetwork['share-url'].replace('[[URL]]', this.getMjAttribute('href'));
    }

    return ['icon-size', 'href', 'src', 'background-color'].reduce(function (r, attr) {
      r[attr] = socialNetwork[attr] || _this.getMjAttribute(attr);

      return r;
    }, {});
  },
  render: function render() {
    var _getSocialAttributes2 = this.getSocialAttributes(),
        src = _getSocialAttributes2.src,
        href = _getSocialAttributes2.href,
        iconSize = _getSocialAttributes2['icon-size'];

    return '\n      <tr>\n        <td ' + this.generateHtmlAttributes({ style: 'td' }) + '>\n          <table\n            ' + this.generateHtmlAttributes({
      border: '0',
      cellpadding: '0',
      cellspacing: '0',
      role: 'presentation',
      style: 'table'
    }) + '\n          >\n            <tr>\n              <td ' + this.generateHtmlAttributes({ style: 'icon' }) + '>\n                <a ' + this.generateHtmlAttributes({
      href: href,
      rel: this.getMjAttribute('rel')
    }) + '>\n                    <img\n                      ' + this.generateHtmlAttributes({
      alt: this.getMjAttribute('alt'),
      height: parseInt(iconSize),
      src: src,
      style: 'img',
      width: parseInt(iconSize)
    }) + '\n                    />\n                  </a>\n                </td>\n                ' + (this.getMjContent() ? '\n                  <td>\n                    <a\n                      ' + this.generateHtmlAttributes({
      href: href,
      style: 'text',
      rel: this.getMjAttribute('rel')
    }) + '>\n                      ' + this.getMjContent() + '\n                    </a>\n                  </td>\n                  ' : '') + '\n              </tr>\n          </table>\n        </td>\n      </tr>\n    ';
  }
});
module.exports = exports['default'];
