// mjml-core/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mjml2html = exports.MJMLValidator = exports.version = exports.mjIncludeRegexp = exports.documentParser = exports.rulesCollection = exports.registerMJRule = exports.formatValidationError = exports.MJMLHeadElements = exports.registerMJHeadElement = exports.helpers = exports.elements = exports.registerMJElement = exports.MJMLRenderer = exports.MJMLElement = undefined;

var _includeExternal = require('./includeExternal');

Object.defineProperty(exports, 'mjIncludeRegexp', {
  enumerable: true,
  get: function get() {
    return _includeExternal.includes;
  }
});

var _MJMLRenderer = require('./MJMLRenderer');

var _MJMLRenderer2 = _interopRequireDefault(_MJMLRenderer);

var _mjmlValidator = require('mjml-validator');

var _mjmlValidator2 = _interopRequireDefault(_mjmlValidator);

var _MJMLElementsCollection = require('./MJMLElementsCollection');

var _MJMLElementsCollection2 = _interopRequireDefault(_MJMLElementsCollection);

var _MJMLHead = require('./MJMLHead');

var _MJMLHead2 = _interopRequireDefault(_MJMLHead);

var _helpers = require('./helpers');

var helpers = _interopRequireWildcard(_helpers);

var _MJMLElement2 = require('./decorators/MJMLElement');

var _MJMLElement3 = _interopRequireDefault(_MJMLElement2);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.MJMLElement = _MJMLElement3.default;
exports.MJMLRenderer = _MJMLRenderer2.default;
exports.registerMJElement = _MJMLElementsCollection.registerMJElement;
exports.elements = _MJMLElementsCollection2.default;
exports.helpers = helpers;
exports.registerMJHeadElement = _MJMLHead.registerMJHeadElement;
exports.MJMLHeadElements = _MJMLHead2.default;
exports.formatValidationError = _mjmlValidator.formatValidationError;
exports.registerMJRule = _mjmlValidator.registerMJRule;
exports.rulesCollection = _mjmlValidator.rulesCollection;
var documentParser = exports.documentParser = function documentParser(content, attributes) {
  var documentParser = require('./parsers/document').default;

  return documentParser(content, attributes);
};
var version = exports.version = function version() {
  return '3.3.3';
};
var MJMLValidator = exports.MJMLValidator = _mjmlValidator2.default;
var mjml2html = exports.mjml2html = function mjml2html(mjml) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new _MJMLRenderer2.default(mjml, options).render();
};
