// mjml-group/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _mjmlCore = require('mjml-core');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var tagName = 'mj-group';
var parentTag = ['mj-section', 'mj-navbar'];
var defaultMJMLDefinition = {
  attributes: {
    'width': null,
    'background-color': null,
    'vertical-align': null
  }
};
var baseStyles = {
  div: {
    verticalAlign: 'top'
  }
};
var postRender = function postRender($) {
  $('.mj-group-outlook-open').each(function () {
    var $parent = $(this).parent();
    var mjGroupBg = $parent.data('mj-group-background');
    var $columnDiv = $(this).next();
    var bgColor = mjGroupBg ? 'bgcolor="' + mjGroupBg + '"' : '';
    var classes = $columnDiv.attr('data-class') ? $columnDiv.attr('data-class').split(' ').map(function (c) {
      return c + '-outlook';
    }).join(' ') : false;
    $columnDiv.removeAttr('data-class');

    $(this).replaceWith(_mjmlCore.helpers.startConditionalTag + '\n      <table ' + bgColor + ' role="presentation" border="0" cellpadding="0" cellspacing="0">\n        <tr>\n          <td style="vertical-align:' + $columnDiv.data('vertical-align') + ';width:' + parseInt($(this).data('width')) + 'px;"' + (classes ? ' class="' + classes + '"' : '') + '>\n      ' + _mjmlCore.helpers.endConditionalTag);

    $parent.removeAttr('data-mj-group-background');
    $columnDiv.removeAttr('data-vertical-align');
  });

  $('.mj-group-outlook-line').each(function () {
    var $columnDiv = $(this).next();
    var classes = $columnDiv.attr('data-class') ? $columnDiv.attr('data-class').split(' ').map(function (c) {
      return c + '-outlook';
    }).join(' ') : false;
    $columnDiv.removeAttr('data-class');

    $(this).replaceWith(_mjmlCore.helpers.startConditionalTag + '\n            </td>\n            <td style="vertical-align:' + $columnDiv.data('vertical-align') + ';width:' + parseInt($(this).data('width')) + 'px;"' + (classes ? ' class="' + classes + '"' : '') + '>\n      ' + _mjmlCore.helpers.endConditionalTag);

    $columnDiv.removeAttr('data-vertical-align');
  });

  $('.mj-group-outlook-close').each(function () {
    $(this).replaceWith(_mjmlCore.helpers.startConditionalTag + '\n            </td>\n          </tr>\n        </table>\n      ' + _mjmlCore.helpers.endConditionalTag);
  });

  return $;
};

var Group = (0, _mjmlCore.MJMLElement)(_class = function (_Component) {
  _inherits(Group, _Component);

  function Group() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Group);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Group.__proto__ || Object.getPrototypeOf(Group)).call.apply(_ref, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Group, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return _mjmlCore.helpers.merge({}, baseStyles, {
        div: {
          background: mjAttribute('background-color'),
          display: 'inline-block',
          verticalAlign: mjAttribute('vertical-align'),
          fontSize: '0px',
          lineHeight: '0px',
          textAlign: 'left',
          width: '100%'
        }
      });
    }
  }, {
    key: 'getGroupClass',
    value: function getGroupClass() {
      var _props = this.props,
          mjAttribute = _props.mjAttribute,
          sibling = _props.sibling;

      var width = mjAttribute('width');
      var parentWidth = this.props.parentWidth || mjAttribute('parentWidth');

      if (width == undefined) {
        return 'mj-column-per-' + parseInt(100 / sibling);
      }

      var _helpers$widthParser = _mjmlCore.helpers.widthParser(width),
          parsedWidth = _helpers$widthParser.width,
          unit = _helpers$widthParser.unit;

      switch (unit) {
        case '%':
          return 'mj-column-per-' + parsedWidth;

        case 'px':
          var percentWidth = parseInt(width) / parentWidth * 100;
          return 'mj-column-per-' + percentWidth;
      }
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var children = this.props.children;


      return children.map(function (child) {
        return _react2.default.cloneElement(child, { mobileWidth: "mobileWidth" });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          mjAttribute = _props2.mjAttribute,
          sibling = _props2.sibling,
          renderWrappedOutlookChildren = _props2.renderWrappedOutlookChildren;

      var width = mjAttribute('width') || 100 / sibling;
      var mjGroupClass = this.getGroupClass();
      var divClasses = (0, _classnames2.default)(mjGroupClass, 'outlook-group-fix', mjAttribute('css-class'));

      return _react2.default.createElement(
        'div',
        {
          className: divClasses,
          'data-class': mjAttribute('css-class'),
          'data-column-width': parseInt(width),
          'data-vertical-align': this.styles.div.verticalAlign,
          'data-mj-group-background': mjAttribute('background-color'),
          style: this.styles.div },
        renderWrappedOutlookChildren(this.renderChildren())
      );
    }
  }]);

  return Group;
}(_react.Component)) || _class;

Group.tagName = tagName;
Group.baseStyles = baseStyles;
Group.postRender = postRender;
Group.parentTag = parentTag;
Group.defaultMJMLDefinition = defaultMJMLDefinition;

exports.default = Group;
