// mjml-location/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _mjmlCore = require('mjml-core');

var _mjmlImage = require('mjml-image');

var _mjmlImage2 = _interopRequireDefault(_mjmlImage);

var _mjmlText = require('mjml-text');

var _mjmlText2 = _interopRequireDefault(_mjmlText);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var tagName = 'mj-location';
var parentTag = ['mj-column', 'mj-hero-content'];
var endingTag = true;
var selfClosingTag = true;
var defaultMJMLDefinition = {
  attributes: {
    'address': null,
    'rel': null,
    'align': null,
    'color': '#3aa7ed',
    'container-background-color': null,
    'font-family': 'Roboto, sans-serif',
    'font-size': '18px',
    'font-weight': '500',
    'img-src': 'http://i.imgur.com/DPCJHhy.png',
    'padding-bottom': null,
    'padding-left': null,
    'padding-right': null,
    'padding-top': null,
    'padding': '10px 25px',
    'vertical-align': null
  }
};

var Location = (0, _mjmlCore.MJMLElement)(_class = function (_Component) {
  _inherits(Location, _Component);

  function Location() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Location);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Location.__proto__ || Object.getPrototypeOf(Location)).call.apply(_ref, [this].concat(args))), _this), _this.styles = _this.getStyles(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Location, [{
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;


      return {
        text: {
          color: mjAttribute('color'),
          textDecoration: 'none'
        }
      };
    }
  }, {
    key: 'getAttributes',
    value: function getAttributes() {
      var _props = this.props,
          mjAttribute = _props.mjAttribute,
          defaultUnit = _props.defaultUnit;


      return {
        text: {
          'columnElement': true,
          'font-family': mjAttribute('font-family'),
          'font-size': defaultUnit(mjAttribute('font-size')),
          'font-weight': mjAttribute('font-weight'),
          'padding': '0px',
          'text-decoration': mjAttribute('text-decoration')
        },
        img: {
          'columnElement': true,
          'padding': '0px',
          'src': mjAttribute('img-src')
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var mjAttribute = this.props.mjAttribute;

      var attrs = this.getAttributes();

      var address = mjAttribute('href') || 'http://maps.google.com/maps?q=' + encodeURIComponent(mjAttribute('address'));
      var text = mjAttribute('text') || mjAttribute('address');

      return _react2.default.createElement(
        'table',
        {
          role: 'presentation',
          cellPadding: '0',
          cellSpacing: '0',
          'data-legacy-border': '0',
          width: '100%' },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(_mjmlImage2.default, _extends({}, attrs.img, {
            href: address,
            rel: mjAttribute('rel') })),
          _react2.default.createElement(
            _mjmlText2.default,
            _extends({}, attrs.text, {
              align: 'center' }),
            _react2.default.createElement(
              'a',
              {
                href: address,
                rel: mjAttribute('rel'),
                style: this.styles.text,
                target: '_blank' },
              text
            )
          )
        )
      );
    }
  }]);

  return Location;
}(_react.Component)) || _class;

Location.tagName = tagName;
Location.parentTag = parentTag;
Location.endingTag = endingTag;
Location.defaultMJMLDefinition = defaultMJMLDefinition;
Location.selfClosingTag = selfClosingTag;

exports.default = Location;
